File: Assets/Typescript~/src/Server/Services/Match/Map/MapService.ts
import { BWEditorConfig } from "@Easy/Core/Shared/Editor/BWEditorConfig";
import { RandomUtil } from "@Easy/Core/Shared/Util/RandomUtil";
import { RunUtil } from "@Easy/Core/Shared/Util/RunUtil";
import { Task } from "@Easy/Core/Shared/Util/Task";
import { WorldAPI } from "@Easy/Core/Shared/VoxelWorld/WorldAPI";
import { DecodeJSON } from "@Easy/Core/Shared/json";
import { Dependency, OnStart, Service } from "@easy-games/flamework-core";
import { ServerSignals } from "Server/ServerSignals";
import { MapLoadEvent } from "Server/Signals/MapLoadEvent";
import { QueueMeta } from "Shared/Queue/QueueMeta";
import { MatchService } from "../MatchService";
import { LoadedMap } from "./LoadedMap";
import { WorldPosition } from "./MapPosition";

@Service({})
export class MapService implements OnStart {
	/** Map for current match. */
	private gameMapId: string | undefined;
	/** Voxel binary file for current map. */
	private voxelBinaryFile: WorldSaveFile | undefined;
	/** Queue meta for current match. */
	private queueMeta: QueueMeta | undefined;
	/** Loaded map instance for current match. */
	private loadedMap: LoadedMap | undefined;
	/** Whether or not map is fully loaded. */
	private mapLoaded = false;

	OnStart(): void {
		Task.Spawn(() => {
			this.queueMeta = Dependency<MatchService>().WaitForQueueReady();

			let mapId = RandomUtil.FromArray(this.queueMeta.maps);
			if (RunUtil.IsEditor()) {
				const gameConfigFile = AssetBridge.Instance.LoadAsset<TextAsset>(
					"Shared/Resources/BWEditorConfig.json",
				);
				const gameConfig = DecodeJSON<BWEditorConfig>(gameConfigFile.text);
				mapId = gameConfig.gameMap;
			}

			this.gameMapId = mapId;
			this.BuildMap(this.gameMapId);
		});
	}

	public BuildMap(mapId: string): void {
		/* Fetch world, load map voxel file and block defines. */
		print("Loading world " + mapId);
		const world = WorldAPI.GetMainWorld();
		if (!world) {
			Debug.LogError("VoxelWorld did not exist. Can't build map.");
			return;
		}
		this.voxelBinaryFile = AssetBridge.Instance.LoadAsset<WorldSaveFile>(`Server/Resources/Worlds/${mapId}.asset`);

		/* Load world. */
		// world.LoadEmptyWorld(blockDefines, "");
		// const grass = GetItemDef(ItemType.GRASS).BlockId;
		// world.WriteVoxelAt(new Vector3(1, 1, 1), grass!);
		world.LoadWorldFromSaveFile(this.voxelBinaryFile);
		/* Parse map objects and finish loading map. */
		/* TEMP: This is to get around memory pinning issue. */
		this.loadedMap = new LoadedMap(mapId, mapId, ["Rascal"]);
		const rawMaps = this.voxelBinaryFile.GetMapObjects();
		for (let i = 0; i < rawMaps.Length; i++) {
			const data = rawMaps.GetValue(i);
			this.loadedMap.AddWorldPositions(data.name, new WorldPosition(data.position, data.rotation));
		}

		world.WaitForFinishedLoading().expect();

		ServerSignals.MapLoad.Fire(new MapLoadEvent(this.loadedMap));
		this.mapLoaded = true;
	}

	/**
	 * Yields until map loads.
	 * @returns The loaded map.
	 */
	public WaitForMapLoaded(): LoadedMap {
		if (this.loadedMap && this.mapLoaded && this.voxelBinaryFile) return this.loadedMap;
		while (!this.loadedMap && !this.mapLoaded && !this.voxelBinaryFile) {
			task.wait(0.1);
		}
		return this.loadedMap!;
	}

	/**
	 * @returns The loaded map.
	 */
	public GetLoadedMap(): LoadedMap | undefined {
		return this.loadedMap;
	}

	/**
	 * @returns The game map.
	 */
	public GetGameMapId(): string | undefined {
		return this.gameMapId;
	}
}

File: Assets/Bundles/Server/Resources/TS/Services/Match/Map/MapService.lua
-- Compiled with unity-ts v2.1.0-75
local _flamework_core = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init")
local Reflect = _flamework_core.Reflect
local Flamework = _flamework_core.Flamework
local Service = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Service
local ServerSignals = require("Server/TS/ServerSignals").ServerSignals
local RandomUtil = require("Shared/TS/Util/RandomUtil").RandomUtil
local Task = require("Shared/TS/Util/Task").Task
local WorldAPI = require("Shared/TS/VoxelWorld/WorldAPI").WorldAPI
local LoadedMap = require("Server/TS/Services/Match/Map/LoadedMap").LoadedMap
local MapService
do
	MapService = setmetatable({}, {
		__tostring = function()
			return "MapService"
		end,
	})
	MapService.__index = MapService
	function MapService.new(...)
		local self = setmetatable({}, MapService)
		return self:constructor(...) or self
	end
	function MapService:constructor()
		self.mapLoaded = false
	end
	function MapService:OnStart()
		Task:Spawn(function()
			self.queueMeta = (Flamework.resolveDependency("Bundles/Server/Services/Match/MatchService@MatchService")):WaitForQueueReady()
			self.gameMap = RandomUtil:FromArray(self.queueMeta.maps)
			self:BuildMap(self.gameMap)
		end)
	end
	function MapService:BuildMap(map)
		-- Fetch world, load map voxel file and block defines.
		print("Loading world " .. map)
		local world = WorldAPI:GetMainWorld()
		self.voxelBinaryFile = AssetBridge:LoadAsset("Server/Resources/Worlds/" .. (map .. ".asset"))
		local blockDefines = AssetBridge:LoadAsset("Shared/Resources/VoxelWorld/BlockDefines.xml")
		-- Load world.
		-- world.LoadEmptyWorld(blockDefines, "");
		-- const grass = GetItemMeta(ItemType.GRASS).BlockId;
		-- world.WriteVoxelAt(new Vector3(1, 1, 1), grass!);
		world:LoadWorldFromVoxelBinaryFile(self.voxelBinaryFile, blockDefines)
		-- Parse map objects and finish loading map.
		-- TEMP: This is to get around memory pinning issue.
		local mapObjects = {}
		local rawMaps = self.voxelBinaryFile:GetMapObjects()
		do
			local i = 0
			local _shouldIncrement = false
			while true do
				if _shouldIncrement then
					i += 1
				else
					_shouldIncrement = true
				end
				if not (i < rawMaps.Length) then
					break
				end
				local data = rawMaps:GetValue(i)
				local _mapObjects = mapObjects
				local _arg0 = {
					name = data.name,
					position = data.position,
					rotation = data.rotation,
				}
				table.insert(_mapObjects, _arg0)
			end
		end
		self.loadedMap = LoadedMap.new(map, mapObjects)
		ServerSignals.MapLoad:fire(self.loadedMap)
		self.mapLoaded = true
	end
	function MapService:WaitForMapLoaded()
		if self.loadedMap and (self.mapLoaded and self.voxelBinaryFile) then
			return self.loadedMap
		end
		while not self.loadedMap and (not self.mapLoaded and not self.voxelBinaryFile) do
			Task:Wait(0.1)
		end
		return self.loadedMap
	end
	function MapService:GetLoadedMap()
		return self.loadedMap
	end
	function MapService:GetGameMap()
		return self.gameMap
	end
end
-- (Flamework) MapService metadata
Reflect.defineMetadata(MapService, "identifier", "Bundles/Server/Services/Match/Map/MapService@MapService")
Reflect.defineMetadata(MapService, "flamework:implements", { "$:flamework@OnStart" })
Reflect.decorate(MapService, "$:flamework@Service", Service, { {} })
return {
	MapService = MapService,
}
-- ----------------------------------
-- ----------------------------------
